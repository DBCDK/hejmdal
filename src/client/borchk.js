var storageAvailable = true;

// Library input
var libraryGroup;
var librariesDropdownContainer;
var libraryInput;
var libraryIdInput;

// Library input buttons (caret/clear)
var dropDownCaret;
var dropDownClear;

// dropdown vars
var allAgencies;
var storedAgencies = [];
var currentlyVisibleAgencies = [];
var currentlySelectedIndex = -1;
var currentSearchValue = '';

// Other form inputs
var userIdInput;

document.addEventListener('DOMContentLoaded', function() {
  // init sorage availability
  storageAvailable = Storage ? true : false;

  // Update library var nodes
  libraryGroup = document.getElementById('library-group');
  libraryInput = document.getElementById('libraryname-input');
  libraryIdInput = document.getElementById('libraryid-input');
  librariesDropdownContainer = document.getElementById(
    'libraries-dropdown-container'
  );

  // Set userId var nodes
  userIdInput = document.getElementById('userid-input');

  // Set clear/close var nodes
  dropDownCaret = document.getElementById('libraries-dropdown-toggle-btn');
  dropDownClear = document.getElementById('clear-libraries-input-btn');

  // Set agencies
  allAgencies = document.getElementsByClassName('agency');

  // Set storedAgencies
  if (storageAvailable) {
    // fetch recetnly selected libraries from localStorage
    storedAgencies = JSON.parse(localStorage.getItem('agencies')) || [];

    // Create recently selected list in dropdown
    initRecentlySelectedLibraries();
  }

  //
  // Event Listeners
  //

  // register focus in the library select inputfield
  libraryInput.addEventListener('focus', function() {
    dropdownTrigger('open');
  });

  // listen on keyevents in the library select field
  libraryInput.addEventListener('keyup', function() {
    // Other KeyPress'
    if (libraryInput.value !== currentSearchValue) {
      currentSearchValue = libraryInput.value;
      initVisibleLibraries();
    }
    initButtonStatus();
    toggleLabelsInDropDown();
  });

  // Handle dropdown navigation keys ESC | ENTER | UP | DOWN | TAB
  libraryInput.addEventListener('keydown', handleKeyEvents);

  // only for not predefined libraries
  if (librariesDropdownContainer) {
    // detect clicks outside of the dropdown - to close the dropdown
    document.addEventListener('mousedown', function(e) {
      if (librariesDropdownContainer.contains(e.target)) {
        return;
      }

      if (!e.target.classList.contains('prevent-body-close-event')) {
        if (librariesDropdownContainer.classList.contains('visible')) {
          dropdownTrigger('close');
        }
      }
    });

    // Set current search value
    currentSearchValue = libraryInput.value;

    // init
    initVisibleLibraries();
  }
});

//
// Functions
//

// triggers/toggles the dropdown
// status = 'open' || 'close' || 'toggle' (default)
function dropdownTrigger(status = 'toggle') {
  if (status === 'open') {
    librariesDropdownContainer.classList.add('visible');
    libraryGroup.classList.add('dropdown-visible');
  } else if (status === 'close') {
    librariesDropdownContainer.classList.remove('visible');
    libraryGroup.classList.remove('dropdown-visible');
  } else {
    librariesDropdownContainer.classList.toggle('visible');
    libraryGroup.classList.toggle('dropdown-visible');
  }

  toggleLabelsInDropDown();
}

// When a library in the dropdown is clicked/selected
// element = element node
function librarySelect(element) {
  // eslint-disable-line no-unused-vars
  var branchName = element.dataset.name;
  var branchId = element.dataset.aid;

  libraryIdInput.value = branchId;
  libraryInput.value = branchName;
  dropdownTrigger('close');
  userIdInput.focus();
  addLibraryToLocalStorage({branchName, branchId});
  initButtonStatus();
}

// Evaluates if the caret or the clear button (in the libraryfield) should be hidden
function initButtonStatus() {
  if (libraryInput.value.length) {
    dropDownCaret.classList.add('hide');
    dropDownClear.classList.remove('hide');
  } else {
    dropDownCaret.classList.remove('hide');
    dropDownClear.classList.add('hide');
  }
}

// Clears the library input field on clear button click
/* eslint-disable no-unused-vars */
function clearLibraryInput() {
  libraryInput.value = '';
  libraryIdInput.value = '';
  currentSearchValue = '';

  initButtonStatus();
  dropdownTrigger('close');
  clearVisibleLibraries();
}
/* eslint-enable no-unused-vars */

// Toggle Field text visibility (type: password || type: tel)
// id = id of the field
/* eslint-disable no-unused-vars */
function toggleFieldVisibility(id) {
  var field = document.getElementById(id);
  var currentType = field.getAttribute('type');
  var newType = currentType === 'password' ? 'tel' : 'password';
  field.setAttribute('type', newType);
}
/* eslint-enable no-unused-vars */

// Toggle labels in dropdown
function toggleLabelsInDropDown() {
  if (currentSearchValue.length > 0) {
    document.getElementById('latest').classList.add('hide');
    document.getElementById('alphabetical').classList.add('hide');
  } else {
    if (storedAgencies.length > 0) {
      document.getElementById('latest').classList.remove('hide');
    }
    document.getElementById('alphabetical').classList.remove('hide');
  }
}

// Evaluate which libraries to be visible in the dropdown
function initVisibleLibraries() {
  currentlyVisibleAgencies = [];

  for (let i = 0; i < allAgencies.length; i++) {
    const item = allAgencies.item(i);
    item.classList.remove('selected');
    let shouldHide =
      allAgencies
        .item(i)
        .textContent.toLowerCase()
        .indexOf(currentSearchValue.toLowerCase()) === -1;
    item.classList.toggle('hide', shouldHide);

    for (let j = 0; j < currentlyVisibleAgencies.length; j++) {
      if (currentlyVisibleAgencies[j].innerText === item.innerText) {
        shouldHide = true;
        item.classList.add('hide');
      }
    }

    if (!shouldHide) {
      currentlyVisibleAgencies.push(item);
      currentlySelectedIndex = -1;
    }
  }
}

// clear visible labraries
function clearVisibleLibraries() {
  currentlyVisibleAgencies = [];
  currentlySelectedIndex = -1;

  for (let i = 0; i < allAgencies.length; i++) {
    const item = allAgencies.item(i);
    item.classList.remove('selected');
    item.classList.remove('hide');
    currentlyVisibleAgencies.push(item);
  }
}

// Creats a list with recently selected libraries in top of the dropdown
function initRecentlySelectedLibraries() {
  clearRecentlySelectedLibraries();
  if (storedAgencies.length === 0 || !librariesDropdownContainer) {
    return;
  }

  const alphabeticalHeader = document.getElementById('alphabetical');
  const librariesDropdown = document.getElementById('libraries-dropdown');

  storedAgencies.forEach(function(agency) {
    const li = document.createElement('li');
    li.classList.add('agency');
    li.classList.add('recent');
    li.setAttribute('data-aid', agency.branchId);
    li.setAttribute('data-name', agency.branchName);
    li.setAttribute('onclick', 'librarySelect(this)');
    li.appendChild(document.createTextNode(agency.branchName));
    librariesDropdown.insertBefore(li, alphabeticalHeader);
  });

  toggleLabelsInDropDown();
}

// Reset the latest selected libraries list in dropdown
function clearRecentlySelectedLibraries() {
  var elements = document.getElementsByClassName('recent');
  var lis = [];

  for (var i = 0; i < elements.length; i++) {
    var li = elements.item(i);
    lis.push(li);
  }

  lis.forEach(function(_li) {
    _li.parentNode.removeChild(_li);
  });
}

// Add library to localStorage (Recently selected list)
function addLibraryToLocalStorage({branchName, branchId}) {
  if (!storageAvailable) {
    // eslint-disable-line no-undefined
    return;
  }

  let indexOfExistingItem = -1;
  storedAgencies.find(function(element, index) {
    if (element.branchId === branchId) {
      indexOfExistingItem = index;
    }
  });

  if (indexOfExistingItem >= 0) {
    storedAgencies.splice(indexOfExistingItem, 1);
  }

  storedAgencies.splice(0, 0, {branchName, branchId});

  if (storedAgencies.length >= 7) {
    storedAgencies.pop();
  }

  localStorage.setItem('agencies', JSON.stringify(storedAgencies));
  initRecentlySelectedLibraries();
}

// Keybord events
function handleKeyEvents(e) {
  if (!e.key && e.keyCode) {
    e.key = parseKeyCode(e.keyCode);
  }
  if (e.key === 'ArrowDown' || e.key === 'ArrowUp') {
    e.preventDefault();
    navigateDropDown(e.key);
  }

  if (e.key === 'Tab' || e.key === 'Enter') {
    selectHighlighted(e);
  }

  if (e.key === 'Escape') {
    escapeWasPressed(e);
  }
}

// find keyName based on keyCode
function parseKeyCode(keyCode) {
  let key = '';

  switch (keyCode) {
    case 40:
      key = 'ArrowDown';
      break;
    case 38:
      key = 'ArrowUp';
      break;
    case 9:
      key = 'Tab';
      break;
    case 13:
      key = 'Enter';
      break;
    case 27:
      key = 'Escape';
      break;
    default:
      break;
  }

  return key;
}

// Navigate up/down in the dropdown
function navigateDropDown(key) {
  if (!librariesDropdownContainer.classList.contains('visible')) {
    return;
  }

  if (currentlyVisibleAgencies.length === 0) {
    return;
  }

  if (key === 'ArrowDown') {
    if (currentlySelectedIndex >= 0) {
      currentlyVisibleAgencies[currentlySelectedIndex].classList.remove(
        'selected'
      );
      currentlySelectedIndex++;

      if (currentlySelectedIndex >= currentlyVisibleAgencies.length) {
        currentlySelectedIndex = 0;
      }
    } else {
      currentlySelectedIndex = 0;
    }
    currentlyVisibleAgencies[currentlySelectedIndex].classList.add('selected');
  } else if (key === 'ArrowUp') {
    if (currentlySelectedIndex >= 0) {
      currentlyVisibleAgencies[currentlySelectedIndex].classList.remove(
        'selected'
      );
      currentlySelectedIndex--;

      if (currentlySelectedIndex <= -1) {
        currentlySelectedIndex = currentlyVisibleAgencies.length - 1;
      }
    } else {
      currentlySelectedIndex = currentlyVisibleAgencies.length - 1;
    }
    currentlyVisibleAgencies[currentlySelectedIndex].classList.add('selected');
  }
}

// selected the highlighted library in the dropdown
function selectHighlighted(e) {
  var currentlySelected = currentlyVisibleAgencies[currentlySelectedIndex];

  if (
    currentlySelected &&
    librariesDropdownContainer.classList.contains('visible')
  ) {
    e.preventDefault();
    librarySelect(currentlySelected);
  } else if (librariesDropdownContainer.classList.contains('visible')) {
    e.preventDefault();
    navigateDropDown('ArrowDown');
  }
}

// close dropdown on escape
function escapeWasPressed(e) {
  if (librariesDropdownContainer.classList.contains('visible')) {
    e.preventDefault();
    librariesDropdownContainer.classList.remove('visible');
  }
}

// client form validtion
/* eslint-disable no-unused-vars */
function loginSubmit() {
  // Get inputfields
  if (librariesDropdownContainer) {
    var libraryName = document.getElementById('libraryname-input');
    var libraryId = document.getElementById('libraryid-input');
  }
  var userId = document.getElementById('userid-input');
  var pin = document.getElementById('pin-input');

  // Get input description text
  var libraryText = document.getElementById('libraryname-input-text');
  var idText = document.getElementById('userid-input-text');
  var pinText = document.getElementById('pin-input-text');

  // Reset error messages
  if (librariesDropdownContainer) {
    resetFieldErrorMessage(libraryName, libraryText);
  }
  resetFieldErrorMessage(userId, idText);
  resetFieldErrorMessage(pin, pinText);

  // Error messages
  var noLibraryMessage = 'Du skal vælge et bibliotek';
  var noIdMessage = 'Du skal angive dit Cpr- eller lånernummer';
  var noPinMessage = 'Du skal angive din 4- eller 5-cifrede bibliotekskode';
  var invalidPinMessage = 'Bibliotekskoden skal være på 4 eller 5 cifre.';

  var valid = true;

  // if no libarary selected
  if (librariesDropdownContainer && !libraryId.value) {
    addFieldErrorMessage(libraryName, libraryText, noLibraryMessage);
    valid = false;
  }

  // if no userid typed
  if (!userId.value) {
    addFieldErrorMessage(userId, idText, noIdMessage);
    valid = false;
  }

  // if no pin typed
  if (!pin.value) {
    addFieldErrorMessage(pin, pinText, noPinMessage);
    valid = false;
  }

  // if pin length is not valid
  if (pin.value) {
    if (pin.value.length < 4 || pin.value.length > 5) {
      addFieldErrorMessage(pin, pinText, invalidPinMessage);
      valid = false;
    }
  }

  // if no error found, form is submit
  if (valid) {
    document.getElementById('borchk-login-form').submit();
  }
}
/* eslint-enable no-unused-vars */

// rests the form errors
function resetFieldErrorMessage(field, text) {
  field.parentNode.classList.remove('input-inValid');
  text.classList.remove('text-inValid');
  text.innerText = text.dataset.text;
}

// adds error to form inputs
function addFieldErrorMessage(field, text, message) {
  field.parentNode.classList.add('input-inValid');
  text.classList.add('text-inValid');
  text.innerText = message;
}
