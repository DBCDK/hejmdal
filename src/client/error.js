/**
 * @file
 * JS functionality related to display of errors
 */

document.addEventListener('DOMContentLoaded', function() {
  if (document.getElementById('error-close-btn')) {
    document
      .getElementById('error-close-btn')
      .addEventListener('click', closeErrorOverlay);
  }

  if (document.getElementById('try-again-btn')) {
    document
      .getElementById('try-again-btn')
      .addEventListener('click', closeErrorOverlay);
  }
});

/**
 * Closes the eroroverlay
 * @param {Event} e
 */
function closeErrorOverlay(e) {
  if (e) {
    // e.preventDefault();
  }
  document
    .getElementById('topbar-container')
    .setAttribute('aria-hidden', 'false');
  document
    .getElementById('content-container')
    .setAttribute('aria-hidden', 'false');
  document
    .getElementById('footer-container')
    .setAttribute('aria-hidden', 'false');
}
