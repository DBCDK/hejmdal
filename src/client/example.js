/* eslint-disable */

const sessionId = 'hejmdal-example-applications';

const defaultState = {
  authorizationURL: `${window.location.origin}/oauth/authorization`,
  tokenURL: `${window.location.origin}/oauth/token`,
  userinfoURL: `${window.location.origin}/userinfo/`,
  redirectUri: `${window.location.origin}/example`,
  path: `/login`,
  clientId: '',
  code: null,
  token: null,
  acces_token: null,
  user: null,
  presel: '',
  agency: '',
  agencyType: null,
  idp: '',
  state: '',
  loginUrl: ''
};

window.hejmdal = Object.assign(
  {},
  defaultState,
  JSON.parse(sessionStorage.getItem(sessionId)) || {}
);

/**
 * OnChange callback for inputfields
 */
window.onChange = function onChange({key, val}) {
  window.hejmdal[key] = val;
  setState();
};

/**
 * OnClick callback for Login-button
 */
window.login = function login() {
  window.location = hejmdal.loginUrl;
};

/**
 * OnClick callback for GetTicket-button
 */
window.getToken = function getToken() {
  const xhr = new XMLHttpRequest();
  const url = `${hejmdal.tokenURL}`;
  xhr.open('POST', url, true);
  xhr.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
  xhr.onload = () => {
    if (xhr.responseText.length) {
      const token = JSON.parse(xhr.responseText);
      hejmdal.token = token;
      if (token && token.access_token) {
        hejmdal.access_token = token.access_token;
      }
      setState();
    }
  };

  xhr.send(
    `grant_type=authorization_code&code=${hejmdal.code}&client_id=${
      hejmdal.clientId
    }&client_secret=secret&redirect_uri=${hejmdal.redirectUri}`
  );
};

/**
 * OnClick callback for getUserinfo-button
 */
window.getUserinfo = function getUserinfo() {
  const xhr = new XMLHttpRequest();
  const url = `${hejmdal.userinfoURL}`;
  xhr.open('GET', url, true);
  xhr.setRequestHeader('Authorization', `Bearer ${hejmdal.access_token}`);
  xhr.onload = () => {
    if (xhr.responseText.length) {
      window.hejmdal.userinfo = JSON.parse(xhr.responseText);
      setState();
    }
  };

  xhr.send();
};

window.logout = function logout() {
  hejmdal = defaultState;
  setState();
  window.location = '/logout';
};

/**
 * Parses the URL for queries
 *
 * @return {{}}
 */
function parseQueryString() {
  const qObj = {};
  const queries = window.location.search.substring(1).split('&');
  queries.forEach(query => {
    const params = query.split('=');
    qObj[params[0]] = params[1];
  });

  return qObj;
}

function getAuthUrl({
  clientId,
  redirectUri,
  presel,
  agency,
  agencyType,
  idp,
  state
}) {
  let url = `${
    window.location.origin
  }/oauth/authorize?response_type=code&client_id=${clientId}&redirect_uri=${redirectUri}`;
  if (presel) {
    url += `&presel=${presel}`;
  }

  if (agency) {
    url += `&agency=${agency}`;
  }

  if (agencyType) {
    url += `&agencytype=${agencyType}`;
  }

  if (idp) {
    url += `&idp=${idp}`;
  }
  if (state) {
    url += `&state=${state}`;
  }

  return url;
}
function getTokenCUrl({tokenURL, redirectUri, code, clientId}) {
  return `curl -X POST ${tokenURL} -d "grant_type=authorization_code&code=${code}&client_id=${clientId}&client_secret=secret&redirect_uri=${redirectUri}"`;
}

function getUserinfoCUrl({userinfoURL, access_token}) {
  return `curl ${userinfoURL} -H "Authorization: Bearer ${access_token}"`;
}

function toString(value) {
  return typeof value !== 'string' ? JSON.stringify(value, null, '  ') : value;
}

function setActiveStep({code, access_token}) {
  const stepValue = access_token ? 'userinfo' : code ? 'code' : 'authorization';
  const steps = document.querySelectorAll('[data-step]');
  steps.forEach(step => {
    if (step.getAttribute('data-step') === stepValue) {
      step.classList.add('active');
    } else {
      step.classList.remove('active');
    }
  });
  const hash = `#${stepValue}`;
  if (window.location.hash !== hash) {
    window.location = hash;
  }
}

function checkRequirements(state) {
  const elements = document.querySelectorAll('[data-requirement]');
  elements.forEach(element => {
    const requirement = element.getAttribute('data-requirement');
    if (state[requirement]) {
      element.classList.add('req-ok');
    } else {
      element.classList.remove('req-ok');
    }
  });
}
function updateBindings(state) {
  const elements = document.querySelectorAll('[data-bind]');
  elements.forEach(element => {
    const requirement = element.getAttribute('data-bind');
    if (state[requirement]) {
      if (typeof element.value !== 'undefined') {
        element.value = toString(state[requirement]);
      } else {
        element.textContent = toString(state[requirement]);
      }
    }
  });
}

/**
 * Updates the UI based on the current state
 */
function setState() {
  hejmdal.loginUrl = getAuthUrl(hejmdal);
  hejmdal.tokenCUrl = getTokenCUrl(hejmdal);
  hejmdal.userinfoCUrl = getUserinfoCUrl(hejmdal);

  updateBindings(hejmdal);
  checkRequirements(hejmdal);
  setActiveStep(hejmdal);

  sessionStorage.setItem(sessionId, JSON.stringify(hejmdal));
}

/**
 * Init method called by the end of HTML rendering
 */
(function init() {
  const queryObj = parseQueryString();
  if (queryObj.code) {
    if (hejmdal.code !== queryObj.code) {
      hejmdal.access_token = null;
      hejmdal.token = null;
      hejmdal.userinfo = null;
      hejmdal.code = queryObj.code;
      hejmdal.state = queryObj.state || '';
    }
  } else {
    hejmdal.initialized = true;
    hejmdal.code = false;
    hejmdal.state = '';
  }

  setState();
})();
