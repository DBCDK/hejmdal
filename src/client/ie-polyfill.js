/* eslint-disable */
// Created by MDN {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find}
if (!Array.prototype.find) {
  Array.prototype.find = function(predicate) {
    if (this === null) {
      throw new TypeError('Array.prototype.find called on null or undefined');
    }
    if (typeof predicate !== 'function') {
      throw new TypeError('predicate must be a function');
    }
    var list = Object(this)
    var length = list.length >>> 0;
    var thisArg = arguments[1];
    var value;

    for (var i = 0; i < length; i++) {
      value = list[i];
      if (predicate.call(thisArg, value, i, list)) {
        return value;
      }
    }
    return undefined;
  };
}

// Source: https://gist.github.com/k-gun/c2ea7c49edf7b757fe9561ba37cb19ca
;(function() {
  // helpers
  var regExp = function(name) {
    return new RegExp('(^| )' + name + '( |$)');
  };
  var forEach = function(list, fn, scope) {
    for (var i = 0; i < list.length; i++) {
      fn.call(scope, list[i]);
    }
  };

  // class list object with basic methods
  function ClassList(element) {
    this.element = element;
  }

  ClassList.prototype = {
    add: function() {
      forEach(arguments, function(name) {
        if (!this.contains(name)) {
          this.element.className += this.element.className.length > 0 ? ' ' + name : name;
        }
      }, this);
    },
    remove: function() {
      forEach(arguments, function(name) {
        this.element.className =
          this.element.className.replace(regExp(name), '');
      }, this);
    },
    toggle: function(name) {
      return this.contains(name)
        ? (this.remove(name), false) : (this.add(name), true);
    },
    contains: function(name) {
      return regExp(name).test(this.element.className);
    },
    // bonus..
    replace: function(oldName, newName) {
      this.remove(oldName), this.add(newName);
    }
  };

  // IE8/9, Safari
  if (!('classList' in Element.prototype)) {
    Object.defineProperty(Element.prototype, 'classList', {
      get: function() {
        return new ClassList(this);
      }
    });
  }

  // replace() support for others
  if (window.DOMTokenList && DOMTokenList.prototype.replace == null) {
    DOMTokenList.prototype.replace = ClassList.prototype.replace;
  }
})();
