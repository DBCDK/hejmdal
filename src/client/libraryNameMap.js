export const libMap = {
  folk: {
    '785100': ['Ålborg'],
    '775100': ['Århus'],
    '737000': ['Næver'],
    '714700': ['Frb'],
    '710100': ['Kbh'],
    '758000': ['Åbenrå', 'Åbenraa', 'Aabenrå'],
    '743000': ['Fåborg-Midtfyn'],
    '716900': ['Høje-Tåstrup'],
    '717300': ['Lyngby-Tårbæk', 'Lyngby-Taarbaek', 'Lyngby-Tårbaek'],
    '718500': ['Taarnby']
  },
  forsk: {
    '820050': ['Ålborg Universitetsbibliotek'],
    '851510': ['Ålborg Universitetshospital. Medicinsk Bibliotek'],
    '873870': ['Ålborg Zoo, Biblioteket'],
    '875460': ['Biblioteket, Ålborg Tekniske Gymnasium'],
    '874280': ['UU-Ålborg. Ungdommens Uddannelsesvejledning, Biblioteket'],
    '874060': ['Århus Gymnasium og HF, Biblioteket'],
    '875130': ['Århus Handelsgymnasium'],
    '875290': ['Århus HF & VUC'],
    '875020': ['Århus Maskinmesterskole. Videncenteret'],
    '874570': ['Århus Statsgymnasium, Biblioteket'],
    '874620': ['ARoS Århus Kunstmuseum, Biblioteket'],
    '820080': ['Biblioteket Arkitektskolen Århus'],
    '861160': ['Erhvervsakademi Århus, biblioteket'],
    '861330': ['Jordbrugets UddannelsesCenter Århus. Biblioteket'],
    '861800': ['SOSU Århus'],
    '800010': ['Det Kongelige Bibliotek'],
    '820010': [
      'Det Kongelige Bibliotek - Aarhus',
      'Det Kongelige Bibliotek - Århus',
      'Det Kgl. Bibliotek - Århus'
    ],
    '850010': ['Det Kongelige Bibliotek - Det Administrative Bibliotek'],
    '810010': [
      'Det Kongelige Bibliotek - København',
      'Det Kgl. Bibliotek - Kbh'
    ],
    '830520': ['Det Kongelige Danske Musikkonservatorium, Biblioteket'],
    '875650': ['Frb museerne'],
    '840350': [
      'Biblioteket for Etnologi, Forhistorisk og Klassisk Arkæologisk. Kbh Universitet, SAXO-Instituttet'
    ],
    '875040': ['Center for Undervisningsmidler, Kbh'],
    '875450': ['Handelsskolen Kbh Nord. Bibliotekerne'],
    '875880': ['Institutbiblioteker ved Kbh Universitet (med begrænset udlån)'],
    '876070': [
      'Institutbiblioteker ved Kbh Universitet (udlånes ikke interurbant)'
    ],
    '820190': ['IT-Universitetet i Kbh, Biblioteket 2E09'],
    '860900': ['Kbhs Erhvervsakademi Bibliotek'],
    '830060': ['Kbhs Professionshøjskoles Bibliotek'],
    '875400': ['Kbhs Stadsarkiv. Håndbogsbiblioteket'],
    '851470': ['Kbhs Universitet. Det Humanistiske Fakultet. Polarbiblioteket'],
    '875210': ['Kbhs VUC. Biblioteket'],
    '860980': ['NEXT Uddannelse Kbh'],
    '861150': [
      'SOPU. Sundhed, omsorg, pædagosik. Kbh & Nordsjælland. Biblioteket'
    ],
    '874400': ['Fåborg Gymnasium, Biblioteket'],
    '872200': ['Høje-Tåstrup Gymnasium, Biblioteket'],
    '874460': ['Borupgård Gymnasium, Studiecenteret'],
    '875610': ['Skovgård Museet'],
    '873630': ['Grenå Gymnasium og HF, Biblioteket']
  }
};
