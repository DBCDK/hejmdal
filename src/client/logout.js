class SingleLogout {
  constructor(container, {clients, returnurl}) {
    this.clients = clients;
    this.returnurl = returnurl;
    this.iframeWrapper = document.createElement('div');
    this.iframeWrapper.classList.add('iframeWrapper');
    this.stateWrapper = document.createElement('div');
    this.stateWrapper.classList.add('stateWrapper');
    container.appendChild(this.stateWrapper);
    container.appendChild(this.iframeWrapper);
  }

  async logoutClients() {
    this.updateState(this.onLoading());
    const responses = await Promise.all(
      this.clients.map(client => this.logoutFrame(client, this.iframeWrapper))
    );
    const isSuccessfull =
      responses.filter(response => response !== true).length === 0;
    if (isSuccessfull) {
      this.updateState(this.onSuccess());
    } else {
      this.updateState(this.onError());
    }
  }

  updateState(state) {
    this.stateWrapper.innerHTML = state;
  }

  onLoading() {
    return '<div class="spinner">Vi er i gang med at logge dig ud</div>';
  }
  onSuccess() {
    if (this.returnurl) {
      window.location = this.returnurl;
    }
    return '<div class="success">Du er nu blevet logget ud af bibliotekslogin</div>';
  }
  onError() {
    return '<div>Det har ikke v√¶ret muligt at logge dig ud af alle dine bibliotekstjenester. Du skal istedet lukke din browser for at logge helt ud.</div>';
  }
  logoutFrame(client, wrapper) {
    const iframeElement = document.createElement('iframe');
    iframeElement.setAttribute('id', client.clientId);
    iframeElement.style.visibility = 'hidden';
    iframeElement.style.position = 'absolute';
    wrapper.appendChild(iframeElement);
    iframeElement.setAttribute('src', client.singleLogoutPath);

    return new Promise((resolve, reject) => {
      iframeElement.addEventListener('load', e => {
        try {
          const document =
            iframeElement.contentDocument || iframe.contentWindow.document;
          const {statusCode = 500} = JSON.parse(document.body.innerHTML);
          if (Number(statusCode) === 200) {
            resolve(true);
          } else {
            resolve(false);
          }
        } catch (e) {
          console.error(e);
          resolve(false);
        }
      });
    });
  }
}

document.addEventListener('DOMContentLoaded', () => {
  const singleLogout = new SingleLogout(
    document.getElementById('singlelogout'),
    window.singleLogoutData
  );
  singleLogout.logoutClients();
});
