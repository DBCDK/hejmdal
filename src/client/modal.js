function onDimmerClick() {
  // close all open modals
  var modals = document.getElementsByClassName('modal');
  for (var i = 0; i < modals.length; i++) {
    modals[i].classList.remove('visible');
  }
  // close dimmer
  var dimmer = document.getElementById('dimmer');
  dimmer.classList.remove('visible');
}

function toggleModal(modal, status = 'toggle') {
  var dimmer = document.getElementById('dimmer');
  var modal = document.getElementById(modal);

  // force open modal
  if (status === 'open') {
    dimmer.classList.add('visible');
    modal.classList.add('visible');
    return;
  }

  // force close modal
  if (status === 'close') {
    dimmer.classList.remove('visible');
    modal.classList.remove('visible');
    return;
  }

  // toggle modal
  dimmer.classList.toggle('visible');
  modal.classList.toggle('visible');
}

document.addEventListener('DOMContentLoaded', function() {
  var body = document.getElementsByTagName('body')[0];
  var modal = document.getElementsByClassName('modal')[0];
  var modalBody = document.getElementById('modal-body');
  var modalHeader = document.getElementById('modal-header');

  // Add onScroll class if modal window is scrolled
  modalBody.addEventListener('scroll', function() {
    if (modalBody.scrollTop > 0) {
      modalHeader.classList.add('onScroll');
    } else {
      modalHeader.classList.remove('onScroll');
    }
  });

  body.addEventListener('keyup', function(e) {
    // if ESC pressed
    if (e.keyCode == 27) {
      if (modal.classList.contains('visible')) {
        toggleModal(modal.id, 'close');
      }
    }
  });
});
