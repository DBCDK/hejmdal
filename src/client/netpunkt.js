// Function to toggle password field visibility
window.toggleFieldVisibility = function toggleFieldVisibility(id) {
  var field = document.getElementById(id);
  var currentType = field.getAttribute('type');
  var newType = currentType === 'password' ? 'text' : 'password';
  field.setAttribute('type', newType);
};

/**
 * Function to validate netpunkt form fields (triggered on form submit)
 *
 */
//
window.netpunktSubmit = function netpunktSubmit(event) {
  // get form fields
  var userId = document.getElementById('user-input');
  var groupId = document.getElementById('group-input');
  var password = document.getElementById('pass-input');

  // Reset error messages
  resetFieldErrorMessage(userId, '');
  resetFieldErrorMessage(groupId, '');
  resetFieldErrorMessage(password, '');

  // Error messages
  var errorRequiredField = 'Feltet skal udfyldes';
  var errorInvalidField = 'Indtast et gyldigt gruppe id';

  // Default state
  var valid = true;

  // if no user is is set
  if (!userId.value) {
    addFieldErrorMessage(userId, errorRequiredField);
    valid = false;
  }

  // If no group id is set
  if (!groupId.value) {
    addFieldErrorMessage(groupId, errorRequiredField);
    valid = false;
  }

  if (!password.value) {
    addFieldErrorMessage(password, errorRequiredField);
    valid = false;
  }

  if (groupId.value) {
    // trimming groupId in case of user has entered leading DK-
    const trimmedGroupId = groupId.value.toLowerCase().replace('dk-', '');
    // if trimmed groupId length is not 6 characters
    if (trimmedGroupId.length !== 6) {
      addFieldErrorMessage(groupId, errorInvalidField);
      valid = false;
    }
  }

  // If errors found, do not submit form
  if (!valid) {
    event.preventDefault();
  }
};

/**
 * Function to reset form fields
 *
 * @param {element} field
 */

// resets the form errors
function resetFieldErrorMessage(field) {
  if (!field) {
    return;
  }

  var text = field.parentNode.nextSibling;

  field.parentNode.classList.remove('input-inValid');
  text.classList.remove('text-inValid');
  text.innerText = '';
}

/**
 * Function to add error (and message) to field
 *
 * @param {element} field
 * @param {string} message (optional)
 */

// adds error to form inputs
function addFieldErrorMessage(field, message = '') {
  if (!field) {
    return;
  }

  var text = field.parentNode.nextSibling;

  field.parentNode.classList.add('input-inValid');
  text.classList.add('text-inValid');
  text.innerText = message;
}
