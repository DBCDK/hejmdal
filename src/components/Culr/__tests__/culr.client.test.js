/**
 * @file
 * Unittesting methods in culr.component
 */

import {createAccount, init} from '../culr.client';

describe('Tests for CULR client. Assert that methods are called with correct params', () => {
  beforeAll(async () => {
    await init(false);
  });
  describe('createAccount', () => {
    it('should call createAccount with correct params', async () => {
      const res = await createAccount({
        userIdType: 'CPR',
        userIdValue: 'user',
        agencyId: '1234',
        municipalityNo: '101'
      });
      expect(res).toMatchSnapshot();
    });
    it('should call createAccount without municipalityNo', async () => {
      const res = await createAccount({
        userIdType: 'CPR',
        userIdValue: 'user',
        agencyId: '1234'
      });
      expect(res).toMatchSnapshot();
    });
  });
});
