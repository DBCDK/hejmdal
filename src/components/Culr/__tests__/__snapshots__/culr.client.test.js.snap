// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for CULR client. Assert that methods are called with correct params createAccount should call createAccount with correct params 1`] = `
Object {
  "agencyId": "1234",
  "authCredentials": Object {
    "groupIdAut": "12345",
    "passwordAut": "password",
    "userIdAut": "test_create",
  },
  "municipalityNo": "101",
  "userCredentials": Object {
    "userIdType": "CPR",
    "userIdValue": "user",
  },
}
`;

exports[`Tests for CULR client. Assert that methods are called with correct params createAccount should call createAccount without municipalityNo 1`] = `
Object {
  "agencyId": "1234",
  "authCredentials": Object {
    "groupIdAut": "12345",
    "passwordAut": "password",
    "userIdAut": "test_create",
  },
  "userCredentials": Object {
    "userIdType": "CPR",
    "userIdValue": "user",
  },
}
`;
